pandas/core/indexes/datetimes.py

```diff
@@ -2269,18 +2269,27 @@ def _generate_regular_range(start, end, periods, offset):
         data = DatetimeIndex._simple_new(data, None, tz=tz)
     else:
         if isinstance(start, Timestamp):
-            start = start.to_pydatetime()
+            start_date = start.to_pydatetime()
+        else:
+            start_date = start
 
         if isinstance(end, Timestamp):
-            end = end.to_pydatetime()
+            end_date = end.to_pydatetime()
+        else:
+            end_date = end
 
-        xdr = generate_range(start=start, end=end,
+        xdr = generate_range(start=start_date, end=end_date,
                              periods=periods, offset=offset)
 
         dates = list(xdr)
         # utc = len(dates) > 0 and dates[0].tzinfo is not None
         data = tools.to_datetime(dates)
 
+        # Add back in the lost nanoseconds
+        if isinstance(start, Timestamp) and isinstance(end, Timestamp):
```

Comment:
seems suspect that you need to do this here, what is the reason?

Reply:
This reason for this was that changing ``to_timestamp(how='end')`` to return a time of 23:59:59.99999999 caused problems with ``asfreq`` because ``asfreq`` creates a new ``DatetimeIndex`` which then later calls ``to_pydatetime()`` to convert the Timestamps to native Python datetime objects. This leads to a loss of nanosecond precision. I think there might be an easier way to fix this and I have pushed a new commit to see if it works.